File: matlab/DWI_weight.m
function DWI_weighted = DWI_weight(DWI, h, a, fod_prev, scaling, q)

[n1, n2, n] = size(DWI);
DWI_weighted = zeros(n1, n2, n);

for k1 = 1:n1
    for k2 = 1:n2
        fod_prev(k1, k2, :) = fod_stand(squeeze(fod_prev(k1, k2, :)));
    end
end

if scaling
    weight_min = zeros(n1, n2);
    for k1 = 1:n1
        for k2 = 1:n2
            weight_cand = [];
            if k2 < n2
                weight_cand = [weight_cand, hellinger_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1, k2+1, :)))];
                %weight_cand = [weight_cand, js_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1, k2+1, :)))];
            end
            if k1 < n1
                weight_cand = [weight_cand, hellinger_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1+1, k2, :)))];
                %weight_cand = [weight_cand, js_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1+1, k2, :)))];
            end
            if k2 > 1
                weight_cand = [weight_cand, hellinger_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1, k2-1, :)))];
               %weight_cand = [weight_cand, js_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1, k2-1, :)))];
            end
            if k1 > 1
               weight_cand = [weight_cand, hellinger_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1-1, k2, :)))];
               %weight_cand = [weight_cand, js_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(k1-1, k2, :)))];
            end
            weight_min(k1, k2) = min(weight_cand); %minimum nearest-neighbor hellinger distance
        end
    end
    %smoothness assumption: MNNHD is small for each voxel
    %for those 100(1-q) percent voxels with large MNNHD, they are not well
    %smoothed; when calculating Kst, the hellinger distance corresponding to
    %MNNHD is decreased to 100q percentile of MNNHDs, and all the other
    %hellinger distances are decreased in proportion
    %for those 100q percent voxels with small MNNHD, they are smoothed enough; 
    %when calculating Kst, the hellinger distance corresponding to
    %MNNHD is increased to 100q percentile of MNNHDs, and all the other
    %hellinger distances are increased in proportion
    %weight_ratio_1 = max(weight_min / quantile(reshape(weight_min, n1*n2, 1), q), 1);
    %weight_ratio_2 = min(weight_min / quantile(reshape(weight_min, n1*n2, 1), q), 1);
    %weight_ratio = weight_ratio_1 .* weight_ratio_2; %same as the next line but is more flexible
    weight_ratio = weight_min / quantile(reshape(weight_min, n1*n2, 1), q);
else
    weight_ratio = ones(n1, n2); %no scaling on hellinger distance
end


for k1 = 1:n1
    for k2 = 1:n2
        weight_loc = zeros(n1, n2);
        weight_st = zeros(n1, n2);
        for l1 = max(1, ceil(k1-h)):min(n1, floor(k1+h))
            for l2 = max(1, ceil(k2-h)):min(n2, floor(k2+h))
                weight_loc(l1, l2) = max(0, 1 - ((k1-l1)^2+(k2-l2)^2) / h^2); %quadratic kernel for location
                %weight_loc(l1, l2) = max(0, 1 - sqrt((k1-l1)^2+(k2-l2)^2) / h); %linear kernel for location
                weight_st(l1, l2) = exp(-a / weight_ratio(k1, k2)^2 * hellinger_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(l1, l2, :)))^2);
                %weight_st(l1, l2) = exp(-a / weight_ratio(k1, k2)^2 * js_dis(squeeze(fod_prev(k1, k2, :)), squeeze(fod_prev(l1, l2, :)))^2);
            end
        end
        weight_st(k1, k2) = 1;
        weight = weight_loc .* weight_st;
        weight = weight / sum(sum(weight));
        for l1 = max(1, ceil(k1-h)):min(n1, floor(k1+h))
            for l2 = max(1, ceil(k2-h)):min(n2, floor(k2+h))
                DWI_weighted(k1, k2, :) = DWI_weighted(k1, k2, :) + weight(l1, l2) * DWI(l1, l2, :);
            end
        end
    end
end

File: DWI_weight.py
import numpy as np
import itertools
import time

# standardize fod to have summation one
def fod_stand(fod):
	fod_st = fod.copy()
	fod_st[fod_st < 0] = 0
	fod_st /= np.sum(fod_st)
	return fod_st

# Hellinger distance between two pdfs f1 and f2
def hellinger_dis(f1, f2):
	return max(np.sqrt(0.5 * np.sum((np.sqrt(f1) - np.sqrt(f2))**2)), 1e-4)

# l2 distance between two pdfs f1 and f2
def l2_dis(f1, f2):
	return max(np.linalg.norm(f1-f2), 1e-4)

# calculate minimum-nearest-neighbor distance for each voxel on 2D or 3D grid
# fod: FODs on 2D or 3D grid
# weight_min: minimum-nearest-neighbor distance on 2D or 3D grid
def MNN_Dist(fod, dis_func):

	n_index = fod.shape[:-1]
	index_list = list(np.ndindex(n_index))
	fod_st = np.zeros(fod.shape)
	weight_min = np.zeros(n_index)

	for k in index_list:
		fod_st[k] = fod_stand(fod[k])

	for k in index_list:
		weight_cand = []
		for i in range(len(n_index)):
			if k[i] < n_index[i]-1:
				k_nb = np.array(k)
				k_nb[i] += 1
				weight_cand.append(dis_func(fod_st[k], fod_st[tuple(k_nb)]))
			if k[i] > 0:
				k_nb = np.array(k)
				k_nb[i] -= 1
				weight_cand.append(dis_func(fod_st[k], fod_st[tuple(k_nb)]))
		weight_min[k] = np.min(np.array(weight_cand))

	return weight_min

# calculate kernel weighted DWI signal at kth voxel
def DWI_Weight(k, DWI, h, b, fod_st, dis_func, weight_ratio, dis_scale):

	n_index = DWI.shape[:-1]
	weight_loc = np.zeros(n_index)
	weight_st = np.zeros(n_index)

	range_low = np.maximum(np.zeros(len(n_index)), np.ceil(np.array(k)-h)).astype(int)
	range_up = np.minimum(n_index, np.floor(np.array(k)+h+1)).astype(int)
	range_list = list(itertools.product(*[range(range_low[i], range_up[i]) for i in range(len(n_index))]))

	for l in range_list:
		weight_loc[l] = max(0, 1 - sum((dis_scale*(np.array(k)-np.array(l)))**2) / h**2)
		if weight_loc[l] > 0:
			weight_st[l] = np.exp(-b / weight_ratio[k]**2 * dis_func(fod_st[k], fod_st[l])**2)
	weight_st[k] = 1
	weight = weight_loc * weight_st
	weight /= np.sum(weight)

	DWI_weight = np.zeros(DWI.shape[-1])
	for l in range_list:
		DWI_weight += weight[l] * DWI[l]

	return DWI_weight